framework/src/multiapps/MultiApp.C

```diff
@@ -1197,19 +1197,25 @@ rankConfig(dof_id_type rank,
   // A "slot" is a group of procs/ranks that are grouped together to run a
   // single (sub)app/sim in parallel.
 
-  auto slot_size = std::max(std::min(nprocs / napps, max_app_procs), min_app_procs);
-  dof_id_type nslots = std::min(nprocs / slot_size, napps);
-  auto leftover_procs = nprocs - nslots * slot_size;
-  auto apps_per_slot = napps / nslots;
-  auto leftover_apps = napps % nslots;
+  const processor_id_type slot_size =
+      std::max(std::min(cast_int<processor_id_type>(nprocs / napps), max_app_procs), min_app_procs);
+  const processor_id_type nslots = std::min(
+      nprocs / slot_size,
+      cast_int<processor_id_type>(std::min(
+          static_cast<dof_id_type>(std::numeric_limits<processor_id_type>::max()), napps)));
+  mooseAssert(nprocs >= (nslots * slot_size),
+              "Ensure that leftover procs is represented by an unsigned type");
+  const processor_id_type leftover_procs = nprocs - nslots * slot_size;
+  const dof_id_type apps_per_slot = napps / nslots;
```

Comment:
why back to dof_id

is auto too dangerous here?

Reply:
I think we should be very explicit here. The number of apps is `dof_id_type`, can be potentially hundreds of billions. The number of processors is much lower than that, I think it is 4 bytes by default so is capped around 4 billion.